### DISCUSSION TITLE
Latex inline and blocks delimiter

### DISCUSSION THREAD

**Discussion by basilgass** [thephpleague/commonmark#702] [Q&A]
Hi,

I would like to disable parsing between LaTeX default delimiters. The math rendering is done using javascript (KaTeX).

The default delimiters are:

- $...$ for inline math
- \(...\) for inline math (alternative mode)
- $$...$$ for block math
- \[...\] for block math (alternative mode)

The delimiters with the dollar signs are working, but:
1. I would need to implement the alternative delimiters.
2. LaTeX use lots of \ (backslash) and other special signs... which might interfer with the markdown parser... So everyting between these delimiters must not be parsed at all !

Is there an easy way to make everything between these delimiters untouched (and also keeping the delimiters) ?

Thanks,
Basil

**Comment by colinodell** [comment]
I think your best bet would be to [implement custom parsers](https://commonmark.thephpleague.com/2.0/customization/overview/#add-custom-syntax-with-parsers) that looks for that alternative syntax and parses them as-is to ensure that other parsers don't handle them.

**Reply by basilgass** [reply]
That's what I thought - I'm trying to make such a custom parser, but I can't make it work... 

In particular, in the `getOpeningCharacter()` and `getClosingCharacter()`, what should I return to detect `\(` and `\)` ? 

As the function is singular, I don't know if I should return `\(` or only `\` ?


Also, for the `getMinLength()`, is it the number of characters of the delimiter or the number of times the returned value from `getOpeningCharacter()` ?


I tried to follow some walkthrough and I modified the "Strikethrough" extension. But actually I had no luck with this...


Thanks for your help :)

**Comment by basilgass** [comment]
Here are the files I made so far, heavily using the strikethrough extension...

LaTeX.php
```php
<?php


namespace App\Scolcours\commonmark\latex;

use League\CommonMark\Node\Inline\AbstractInline;
use League\CommonMark\Node\Inline\DelimitedInterface;

final class LaTeX extends AbstractInline implements DelimitedInterface
{
	private string $delimiter;

	public function __construct(string $delimiter = '\\(')
	{
		parent::__construct();

		$this->delimiter = $delimiter;
	}

	public function getOpeningDelimiter(): string
	{
		return '\\(';
	}

	public function getClosingDelimiter(): string
	{
		return '\\)';
	}
}
```

LaTeXinlineDelimiterProcessor.php
```php
<?php


namespace App\Scolcours\commonmark\latex;
	use League\CommonMark\Delimiter\DelimiterInterface;
	use League\CommonMark\Delimiter\Processor\DelimiterProcessorInterface;
	use League\CommonMark\Node\Inline\AbstractStringContainer;

final class LaTeXinlineDelimiterProcessor implements DelimiterProcessorInterface
{
	public function getOpeningCharacter(): string
	{
		return '\\(';
	}

	public function getClosingCharacter(): string
	{
		return '\\)';
	}

	public function getMinLength(): int
	{
		return 1;
	}

	public function getDelimiterUse(DelimiterInterface $opener, DelimiterInterface $closer): int
	{
		$min = \min($opener->getLength(), $closer->getLength());

		return $min >= 1 ? $min : 0;
	}

	public function process(AbstractStringContainer $opener, AbstractStringContainer $closer, int $delimiterUse): void
	{
		$latex = new LaTeX();

		$tmp = $opener->next();
		while ($tmp !== null && $tmp !== $closer) {
			$next = $tmp->next();
			$latex->appendChild($tmp);
			$tmp = $next;
		}

		$opener->insertAfter($latex);
	}
}
```

LaTeXRenderer.php
```php
<?php


namespace App\Scolcours\commonmark\latex;


	use League\CommonMark\Node\Node;
	use League\CommonMark\Renderer\ChildNodeRendererInterface;
	use League\CommonMark\Renderer\NodeRendererInterface;
	use League\CommonMark\Util\HtmlElement;
	use League\CommonMark\Xml\XmlNodeRendererInterface;

final class LaTeXRenderer implements NodeRendererInterface, XmlNodeRendererInterface
{
	/**
	 * @param LaTeX $node
	 *
	 * {@inheritDoc}
	 *
	 * @psalm-suppress MoreSpecificImplementedParamType
	 */
	public function render(Node $node, ChildNodeRendererInterface $childRenderer): \Stringable
	{
		LaTeX::assertInstanceOf($node);

		return new HtmlElement('div',
			$node->data->get('attributes'),
			$childRenderer->renderNodes($node->children())
		);
	}

	public function getXmlTagName(Node $node): string
	{
		return 'math';
	}

	/**
	 * {@inheritDoc}
	 */
	public function getXmlAttributes(Node $node): array
	{
		return [];
	}
}
```

and LaTeXExtension.php
```php
<?php


namespace App\Scolcours\commonmark\latex;


use League\CommonMark\Environment\EnvironmentBuilderInterface;
use League\CommonMark\Extension\ExtensionInterface;

final class LaTeXExtension implements ExtensionInterface {

	public function register( EnvironmentBuilderInterface  $environment ): void {
		$environment->addDelimiterProcessor( new LaTeXinlineDelimiterProcessor() );
		$environment->addRenderer(LaTeX::class, new LaTeXRenderer());
	}

}
```
